strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/event/artifact/MavenArtifactDownloadedEventListener.java

```diff
@@ -0,0 +1,147 @@
+import org.carlspring.strongbox.providers.layout.LayoutProvider;
+import org.carlspring.strongbox.providers.layout.LayoutProviderRegistry;
+import org.carlspring.strongbox.providers.repository.proxied.LocalStorageProxyRepositoryArtifactResolver;
+import org.carlspring.strongbox.providers.repository.proxied.ProxyRepositoryArtifactResolver;
+import org.carlspring.strongbox.providers.repository.proxied.SimpleProxyRepositoryArtifactResolver;
+import org.carlspring.strongbox.services.ArtifactMetadataService;
+import org.carlspring.strongbox.services.ArtifactResolutionService;
+import org.carlspring.strongbox.storage.repository.Repository;
+import org.carlspring.strongbox.storage.repository.RepositoryLayoutEnum;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.NoSuchFileException;
+import java.security.NoSuchAlgorithmException;
+
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.repository.metadata.Metadata;
+import org.apache.maven.index.locator.Locator;
+import org.apache.maven.index.locator.MetadataLocator;
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Przemyslaw Fusik
+ */
+@Component
+public class MavenArtifactDownloadedEventListener
+        implements ArtifactEventListener
+{
+
+    private static final Logger logger = LoggerFactory.getLogger(MavenArtifactDownloadedEventListener.class);
+
+    @Inject
+    private ConfigurationManager configurationManager;
+
+    @Inject
+    private ArtifactResolutionService artifactResolutionService;
+
+    @Inject
+    private ArtifactMetadataService artifactMetadataService;
+
+    @Inject
+    @SimpleProxyRepositoryArtifactResolver.SimpleProxyRepositoryArtifactResolverQualifier
+    private ProxyRepositoryArtifactResolver simpleProxyRepositoryArtifactResolver;
+
+    @Inject
+    @LocalStorageProxyRepositoryArtifactResolver.LocalStorageProxyRepositoryArtifactResolverQualifier
+    private ProxyRepositoryArtifactResolver localStorageProxyRepositoryArtifactResolver;
+
+    @Inject
+    private LayoutProviderRegistry layoutProviderRegistry;
+
+    private Locator metadataLocator = new MetadataLocator();
+
+    @Override
+    public void handle(ArtifactEvent event)
+    {
+        if (event.getType() != ArtifactEventTypeEnum.EVENT_ARTIFACT_FILE_DOWNLOADED.getType())
+        {
+            return;
+        }
+
+        Repository repository = getRepository(event);
+
+        if (!RepositoryLayoutEnum.MAVEN_2.getLayout().equals(repository.getLayout()))
+        {
+            return;
+        }
+
+        resolveArtifactMetadataFile(event);
+    }
+
+
+    private void resolveArtifactMetadataFile(ArtifactEvent event)
+    {
+        try
```

Comment:
I'm sorry, but I'm not quite sure I'm following your code very well...

What's going on over here? What are you trying to locate? Is this a check to see, if the artifact is on the file system?

This should be done via the `LayoutProvider` abstraction, because it handles the file system in a generic way (making it easier to implement Amazon S3 and Google Cloud Storage support later).

Reply:
What does `resolveArtifactMetadataFile` method do:
1. find out if there is an existing `maven-metadata.xml` file for downloaded artifact
2. if there is no such file in the file system, download `maven-metadata.xml` file from the remote repository and store it in the file system, following SB-562 requirements
3. if there is such file in the local file system, download `maven-metadata.xml` and merge it with the existing  `maven-metadata.xml` file in the file system

> This should be done via the LayoutProvider abstraction
if you mean checking  if the artifact is on the file system, OK, I can implement it using `LayoutProvider`. Please confirm.

Reply:
On the other hand, this is strange a little bit. Why do you want to use `LayoutProvider` abstraction if we want it to be working in the `Maven` case ? `maven-metadata.xml` is strict `Maven` thing, generally we are trying to locate `maven-metadata.xml` file, not an artifact in general. Can't we use maven-related tools to locate  `maven-metadata.xml` file ?

Reply:
Thanks for your clear explanation!

Reply:
...And now, regarding your questions -- as explained by @sbespalov , we should try to use `Path` as much as possible. I am not 100% sure, if the Maven Metadata classes that you're referring to can work with `Path`-s. If they can, then very well -- go ahead and use them, but I suspect they actually don't.

On the other hand, in the project we have `AbstractMavenArtifactLocatorOperation`, which I have feeling (off the top of my head) will cover what you need.

Reply:
To resolve the metadata file on the artifactId level I'll have to use below construction:

```java
Repository repository = getRepository(event);
LayoutProvider layoutProvider = layoutProviderRegistry.getProvider(repository.getLayout());
RepositoryPath artifactPath = layoutProvider.resolve(repository).resolve(event.getPath());
Path metadataPath = artifactPath.getParent().getParent().resolve("maven-metadata.xml");
```